transcendence-app/src/user/user.service.ts

```diff
@@ -53,4 +59,78 @@ export class UserService {
   updateUser(userId: string, updateUserDto: UpdateUserDto) {
     return this.userRepository.updateById(userId, updateUserDto);
   }
+
+  private async addAvatarAndUpdateUser(
+    user: User,
+    newAvatarFileDto: LocalFileDto,
+  ): Promise<StreamableFile | null> {
+    const updatedUser = await this.userRepository.addAvatarAndUpdateUser(
+      { id: uuidv4(), createdAt: new Date(Date.now()), ...newAvatarFileDto },
+      user,
+    );
+    if (!updatedUser) {
+      this.localFileService.deleteFileData(newAvatarFileDto.path);
+      return null;
+    }
+    return this.streamAvatarData(newAvatarFileDto);
+  }
+
+  private async updateAvatar(
+    avatarId: string,
+    newAvatarFileDto: LocalFileDto,
+  ): Promise<StreamableFile | null> {
+    const avatarFile = await this.localFileService.getFileById(avatarId);
+    if (!avatarFile) {
+      this.localFileService.deleteFileData(newAvatarFileDto.path);
+      return null;
+    }
+
+    const updatedAvatarFile = await this.localFileService.updateFileById(
+      avatarFile.id,
+      newAvatarFileDto,
+    );
+    if (!updatedAvatarFile) {
+      this.localFileService.deleteFileData(newAvatarFileDto.path);
+      return null;
+    }
+    this.localFileService.deleteFileData(avatarFile.path);
+    return this.streamAvatarData(updatedAvatarFile);
+  }
+
+  async addAvatar(
+    user: User,
+    newAvatarFileDto: LocalFileDto,
+  ): Promise<StreamableFile | null> {
+    if (user.avatarId === null) {
+      return this.addAvatarAndUpdateUser(user, newAvatarFileDto);
+    }
+
+    return this.updateAvatar(user.avatarId, newAvatarFileDto);
+  }
+
+  private streamAvatarData(fileDto: LocalFileDto): StreamableFile {
+    const stream = createReadStream(fileDto.path);
+
+    return new StreamableFile(stream, {
+      type: fileDto.mimetype,
+      disposition: `inline; filename="${fileDto.filename}"`,
+      length: fileDto.size,
+    });
+  }
+
+  async validateAvatarType(path: string): Promise<boolean | undefined> {
+    /**
+     * Import 'file-type' ES-Module in CommonJS Node.js module
+     */
+    const { fileTypeFromFile } = await (eval('import("file-type")') as Promise<
```

Comment:
>Warning: Executing JavaScript from a string is an enormous security risk. It is far too easy for a bad actor to run arbitrary code when you use eval(). See Never use eval()! , below.

[Here.](https://developer.mozilla.org/en-US/docs/web/javascript/reference/global_objects/eval#never_use_eval!)

Reply:
The `file-type` library uses ES modules, and the NestJS compiles into CommonJS modules, which converts the `import` syntax to `require`.

We use `eval` to import the library dynamically at runtime.

Reply:
Maybe we could use something else :thinking:
I will do some research.

Reply:
Fixed, now we use `Function()` instead of `eval()`. Thank you :rocket: